pyomo/contrib/solver/solvers/highs.py

```diff
@@ -518,17 +637,36 @@ def _set_objective(self, obj):
                         expr=coef,
                         highs=self._solver_model,
                     )
-                    self._objective_helpers.append(mutable_objective_coef)
+                    mutable_linear_coefficients.append(mutable_objective_coef)
 
-            self._solver_model.changeObjectiveOffset(value(repn.constant))
-            if not is_constant(repn.constant):
-                mutable_objective_offset = _MutableObjectiveOffset(
-                    expr=repn.constant, highs=self._solver_model
-                )
-                self._objective_helpers.append(mutable_objective_offset)
+            mutable_constant = _MutableObjectiveOffset(
+                expr=repn.constant, highs=self._solver_model
+            )
+
+            if repn.quadratic_vars and len(repn.quadratic_vars) > 0:
+                for ndx, (v1, v2) in enumerate(repn.quadratic_vars):
+                    v1_id = id(v1)
+                    v2_id = id(v2)
+                    v1_ndx = self._pyomo_var_to_solver_var_map[v1_id]
+                    v2_ndx = self._pyomo_var_to_solver_var_map[v2_id]
+
+                    coef = repn.quadratic_coefs[ndx]
+
+                    mutable_quadratic_coefficients.append(
+                        _MutableQuadraticCoefficient(
+                            expr=coef, row_idx=v1_ndx, col_idx=v2_ndx
```

Comment:
I don't think you should store the variable indices here. If a variable gets removed from the model, these indices will become incorrect. Note how the `_MutableObjectiveCoefficient` stores the variable id and the `_pyomo_var_to_solver_var_map`. That way, it can get the correct index at the time `update` is called.

Reply:
I realize this is very convoluted, but I'm not sure how to avoid it.

Reply:
It would be great to add a test for this scenario.

Reply:
Ah yes you're right, let me add one.

Reply:
I have changed the code such that the dictionary index is the id of variable 1 and 2, I think that should address your concern. These indices only get turned to row/col indices when we build the hessian. I am not sold on the Coefficient class having the map itself, it's definitely a model property in my mind.